### ISSUE TITLE
HandleMessagesInsideTransactionScope is ignored when using multiple bus instances

### ISSUE THREAD

**Issue by summerson1985** [rebus-org/Rebus#512]
Hi!

In my Project.Core library I set up a bus in UseMsmqAsOneWayClient mode to be able to send messages from web api, wcf or whatever context.

In my Project.Processor host I have a handler:
  ```
  internal class UpdateCompanyHandler : IHandleMessages<UpdateCompany> {
        private readonly IBus _thisIsOneWayBus;

        public UpdateCompanyHandler(IBus thisIsOneWayBus) {
            _thisIsOneWayBus = thisIsOneWayBus;
        }

        public async Task Handle(UpdateCompany message) {
            Console.WriteLine("Entering UpdateCompany handler for: {0}", message.CompanyId);

            Console.WriteLine("Sending update status: InProcess");
            await _thisIsOneWayBus.Send(new UpdateStatus { CompanyId = message.CompanyId, Status = "InProcess" });
            Thread.Sleep(TimeSpan.FromSeconds(2));

            Console.WriteLine("Sending update status: Done");
            await _thisIsOneWayBus.Send(new UpdateStatus { CompanyId = message.CompanyId, Status = "Done" });
            Thread.Sleep(TimeSpan.FromSeconds(2));

            Console.WriteLine("Exiting UpdateCompany handler");
        }
    }
```
Project.Processor is configured as following:
```
            _rebusConfigurer = Configure.With(_structureMapAdapter)
                .Logging(c => c.None())
                .Options(o => {
                    o.SetNumberOfWorkers(numberOfWorkers);
                    o.SetMaxParallelism(maxParallelism);
                    o.SimpleRetryStrategy(maxDeliveryAttempts: 1);
                    o.HandleMessagesInsideTransactionScope();
                })
                .Transport(t => t.UseMsmq("Processor"))
```

The problem is that HandleMessagesInsideTransactionScope is being ignored and messages are being sent from within UpdateCompanyHandler before UpdateCompanyHandler handler is exited and transaction commits.
Sample output:

```
Entering UpdateCompany handler for: 123
Sending update status: InProcess
Received status update: Company id 123, status: InProcess
Sending update status: Done
Received status update: Company id 123, status: Done
Exiting UpdateCompany handler
```

You can see "Received status update" being printed by UpdateStatusHandler, e.g. UpdateStatus messages are being sent and appear on the queue before completion of transaction in UpdateCompanyHandler.

This is related to Rebus and Rebus.TransactionScope v.0.99.73

**Comment by mookid8000** [comment]
Hmm... this behavior is a consequence of fixing [this issue](https://github.com/rebus-org/Rebus/issues/495) in which weird behavior was experienced because one bus instance could enlist in another bus' transaction context.

The fix is implemented by only enlisting in the transaction context if it belongs to the bus trying to do the send.

In your case, the receiving bus will have a transaction context going on in the message handler, but since you are using a one-way client to do a send (i.e. a separate bus instance), it doesn't get to enlist.

In a way I would say that this is the expected behavior. If I were you, I would simply ensure that the `IBus` that gets injected into handlers is the bus instance that received the message being handled.

**Comment by summerson1985** [comment]
Thanks for your reply.

> I would simply ensure that the IBus that gets injected into handlers is the bus instance that received the message being handled.

This is tricky to achieve as we host multiple IBus instances in the same host to be able to give proper priorities to some messages. Dependency injection needs to be instructed which of IBus instances it needs to inject based on some message context.

**Comment by mookid8000** [comment]
> (...) to be able to give proper priorities to some messages

How do you do that by selectively injecting one-way clients?

**Comment by summerson1985** [comment]
Our Project.Core depends on IBus to be able to send commands from various places. So Project.Core registers this as a singleton of one-way client for its needs. This allows to send commands from any business logic place so to speak.

```
    internal class OneWayClient : IOneWayClient
    {
        private readonly IBus _serviceBus;

        public OneWayClient()
        {
            _serviceBus = Configure.With(new BuiltinHandlerActivator())
                .Logging(l => l.None())
                .Transport(t => t.UseMsmqAsOneWayClient())
                .Start();
        }

        public void Send(string queueName, object message)
        {
            _serviceBus.Advanced.Routing.Send(queueName, message).Wait();
        }
    }
```
The we register one-way-client in Project.Core structuremap set-up like this:
`            For<IOneWayClient>().Use<OneWayClient>().Singleton();`

Then we have do a custom message routing to a named msmqueue:
```
        protected readonly IOneWayClient _oneWayClient;

        protected AdapterBase(IOneWayClient oneWayClient)
        {
            _oneWayClient = oneWayClient;
        }

        public string IncomingQueue
        {
            get
            {
                return string.Format("host.{0}.incoming", GetType().Name);
            }
        }

		public void SomeBusinessLogicMethod() {
			var message = new MyCommand();
            OneWayClient.Send(IncomingQueue, message);
		}
```

Now when Project.Host starts it spins up multiple IBus instances to have a dedicated input queue for each adapter it supports. As we need a number of messages for adapter A not to affect adapter B we decided to split those in separate queues even though message type is the same for both:
```
        public void StartBuses()
        {
            var supportedAdapters = _adapterService.GetSupportedAdapters();
            foreach (var supportedAdapter in supportedAdapters.AdapterDetails)
            {
                var queueName = supportedAdapter.IncomingQueue;
                var bus = Configure.With(_structureMapContainerAdapter)
                    .Options(o =>
                    {
                        o.HandleMessagesInsideTransactionScope();
                    })
                    .Logging(l => l.None())
                    .Transport(t => t.UseMsmq(queueName))
                    .Start();

                _runningBuses.Add(queueName, bus);
            }
        }
```

To summarize Project.Core has one instance of one-way client whilst Project.Host spins up multiple buses to pick messages from queues.

**Comment by mookid8000** [comment]
I am not sure it is going to be resolved by letting a Rebus bus instance enlist in another bus instance's transaction context, which was what was happening before. This "feature" proposed lots of problems, because of the underlying assumption that the two buses were able to communicate on the same "network".

As an example on where things would go wrong consider two Rebus buses configured to use separate Azure Service Bus namespaces... In this case, an `await bus.Send(..)` in a handler of the first bus instance executed on the second bus instance would end up getting sent on the first bus instance's network – and this is clearly NOT what you would expect 😁 

When I thought (a lot) about this problem, I came to the conclusion that _enlisting in another bus' transaction context was wrong for many reasons_, mainly because it would be very hard to support it in a way that would work reliably with any combination of transports. On the other hand, disallowing this was fairly simple – and it would even be possible to do some additional work to effectively achieve the previous behavior.

If you absolutely want to delay sending the messages until the message handler has finished executing, I can help you write an implementation of `IOneWayClient` that enlists its send actions in the ongoing transaction context – this was the previous behavior can be had.

**Comment by summerson1985** [comment]
Good day again!
I am thinking of if I could locate the proper bus instance which is currently handling the message and use that one to send and publish it would pick the transaction scope then.
Could you please advise on how I can do this?

`            var allBusses = IoC.Container.GetAllInstances<IBus>();
            foreach (var bus in allBusses)
            {
              if (bus.IsCurrentlyHandlingMessage()) {
                bus.Send(); 
              }
            }
`

**Comment by mookid8000** [comment]
> However it is not so when there are multiple busses spinned up ine the same process. 

That is not true. Each bus is transactional with respect to itself.

The only case where it is not so, is when a bus instance uses another bus instance to send/publish. In other words, when a bus instance receives a message and uses ANOTHER bus instance to send/publish, THEN the message is sent/published immediately.

> I basically need to pick the proper bus among all registered in order it to obey the behavior on Wiki page..

Ummm.... no! If you let your IoC container inject `IBus` into your message handler, that instance should be the same as the one that received the message. 

What you should do is this:

* Realize that your multiple buses are logically separate applications
* Keep an IoC container instance for each application
* Have one single `IBus` registered in each IoC container

> So I see an issue with multiple buses don't follow the described behavior.

I will go and edit the transactions page to clarify that the behavior described does not include other bus instances.

**Comment by summerson1985** [comment]
Great advice, Mogens!

It took some effort to twist my head around this approach.
Works really well! I ended up having number of containers eqal to number of buses. Each container has a single IBus instance registered.

Tak and have a great evening
